design-documents/graph-ql/extensibility.md

```diff
@@ -1,46 +1,127 @@
+- _Arguments_ - [Spec](http://spec.graphql.org/June2018/#sec-Field-Arguments) | [Docs](https://graphql.org/learn/schema/#arguments)
+- _Input Object_ - [Spec](http://spec.graphql.org/June2018/#sec-Input-Object-Values) | [Docs](https://graphql.org/learn/schema/#input-types)
 
+## Magento GraphQL Vocabulary
 
-**Solution: Wrappers for output and merger for arguments**
+- _Output Object_: An _Object Type_ used as the response type for a mutation (primarily to support extensibility)
 
-Wrappers for output type along with merging capabilities for arguments can solve extensibility and deprecation issues.
 
-```$graphqls
+## Extensibility Requirements
+
+It should be possible to make the following changes to a schema, _without_ introducing breaking changes:
+
+- Add a new root query (backwards-compatible)
+- Add a new mutation (backwards-compatible)
+- Add optional arguments to a query/mutation (some backwards-compat risks)
+- Add extra data to the output of a mutation (backwards-compatible)
+
+The following patterns should be followed to ensure our schema remains extensible, with _minimal_ (ideally no) breaking changes._
+
+## Backwards-Compatible GraphQL Schema Development Patterns
+
+### Mutation Output Objects
+All _mutations_ should return an "Output Object," rather than some concrete type. An "Output Object" is just an Object Type with the specific purpose of returning data _and_ metadata related to a mutation.
+
+It is not possible for a client to send a mutation _and_ separate root queries in the same request. Because of this, it's critical that the output of a mutation be able to add more data over time, as client needs expand.
+
+#### Example - Bad
+```graphql
 type Mutation {
-    generateCustomerToken(
-            email: String!,
-            password: String!
-    ): GenerateCustomerTokenOutput!
+    createFoo: Foo
 }
+```
 
-type GenerateCustomerTokenOutput {
-    token: String!
+#### Example - Good
+```graphql
+type Mutation {
+    createFoo: FooOutput
+}
+
+type ExampleOutput {
+    # Extensions (and core) can extend with more fields at a later date
+    foo: Foo
 }
 ```
 
-With such schema it is possible to extend the list of arguments (not reduce, however). For example, if system integrator got a new requirement to enable multi-factor authentication, the schema can be extended from 3rd party module to support this requirement as follows. All arguments from different modules will be merged and the resulting schema will contain all of them.
+### Multiple Arguments
 
-```$graphqls
-type Mutation {
-    generateCustomerToken(
-            email: String!,
-            password: String!,
-            multi_factor_auth_token: String!
-    ): GenerateCustomerTokenOutput!
+If a query or mutation accepts (or will likely accept) > 1 argument, an Input Object should be used instead, and given the argument name `input`.
```

Comment:
I would say we must always use `input` because we never know how Magento is/will be extended.

Reply:
I think it's very likely you're right here. I have a tendency to try and leave a little wiggle room 😄

Do you want to make any exception for the case of looking something up by ID? I think in the case of a field like `Query.productByID` (as an example) it might be nice to _just_ take the `ID`, and if folks want to add arguments to a unary field, they just have to introduce a new query instead.

If you think this isn's worth it, happy to go forward with your suggested change.

Reply:
Inability to extend is the reason why I don't like `byId` methods. I would use `input` everywhere. Simper rule - less bugs.

Reply:
> Inability to extend is the reason why I don't like byId methods

Does every query need to be infinitely extensible? Can you describe a scenario where a `Query.fooByID` field would need to have its arguments extended, and where that would make more sense than adding a new query?

I think, for single entity lookups (not queries for collections) there aren't many scenarios where it would make sense to take more than an `ID` as input. Most other arguments you'd add to that method would likely change the behavior as well (breaking change).

Please correct me if I'm wrong, but it seems like a lot of the reasons to add more arguments to a field might be a sign that, for simple cases like this, the field is trying to do too much.	

There's also the question of naming - anytime we use a field like `Query.product` instead of `Query.productByID`, we're using a more generic name that's likely to be overloaded later for either uses, leading to things like deprecations and `_v2` fields. Explicit names and granular fields in this case (imo) kind of force the query to be a bit more focused.

> Simper rule - less bugs.

Isn't it the opposite in this case? Adding an argument to an existing field adds branching to a resolver, and a developer needs to take great care to make sure that logic does not impact any existing queries.

Reply:
I don't have strong opinion against `byId`, but would really like to have `input` mandatory.

Reply:
We can do that 👍  Will make changes